src/vs/workbench/parts/preferences/browser/preferencesEditor.ts

```diff
@@ -405,19 +425,46 @@ class PreferencesRenderers extends Disposable {
 		this._editablePreferencesRenderer = editableSettingsRenderer;
 	}
 
-	public filterPreferences(filter: string): number {
-		const defaultPreferencesFilterResult = this._filterPreferences(filter, this._defaultPreferencesRenderer);
-		const editablePreferencesFilterResult = this._filterPreferences(filter, this._editablePreferencesRenderer);
+	public filterPreferences(filter: string, searchProvider: PreferencesSearchProvider): TPromise<number> {
+		if (this._filtersInProgress) {
+			// Resolved/rejected promises have no .cancel()
+			this._filtersInProgress.forEach(p => p.cancel && p.cancel());
+		}
 
-		const defaultPreferencesFilteredGroups = defaultPreferencesFilterResult ? defaultPreferencesFilterResult.filteredGroups : this._getAllPreferences(this._defaultPreferencesRenderer);
-		const editablePreferencesFilteredGroups = editablePreferencesFilterResult ? editablePreferencesFilterResult.filteredGroups : this._getAllPreferences(this._editablePreferencesRenderer);
-		const consolidatedSettings = this._consolidateSettings(editablePreferencesFilteredGroups, defaultPreferencesFilteredGroups);
-		this._settingsNavigator = new SettingsNavigator(filter ? consolidatedSettings : []);
+		if (filter) {
+			const searchModel = searchProvider.startSearch(filter);
+			this._filtersInProgress = [
+				this._filterPreferences(filter, searchModel, this._defaultPreferencesRenderer),
+				this._filterPreferences(filter, searchModel, this._editablePreferencesRenderer)];
+		} else {
+			this._filtersInProgress = [TPromise.wrap(null), TPromise.wrap(null)];
+		}
 
-		return consolidatedSettings.length;
+		return TPromise.join<IFilterResult>(this._filtersInProgress).then(filterResults => {
+			this._filtersInProgress = null;
+			const defaultPreferencesFilterResult = filterResults[0];
+			const editablePreferencesFilterResult = filterResults[1];
+
+			const defaultPreferencesFilteredGroups = defaultPreferencesFilterResult ? defaultPreferencesFilterResult.filteredGroups : this._getAllPreferences(this._defaultPreferencesRenderer);
+			const editablePreferencesFilteredGroups = editablePreferencesFilterResult ? editablePreferencesFilterResult.filteredGroups : this._getAllPreferences(this._editablePreferencesRenderer);
+			const consolidatedSettings = this._consolidateSettings(editablePreferencesFilteredGroups, defaultPreferencesFilteredGroups);
+
+			if (defaultPreferencesFilterResult && defaultPreferencesFilterResult.remoteResult) {
+				// Disable navigation for remote settings search
```

Comment:
I would not like to expose if results are remote or not. Both the results provider could return results using common API. The editor and renderers just depends on this results API.

Reply:
They are rendered differently, so there needs to be some difference.  Could call it "most relevant" results or something like that.